Documentation/contributing/testing/bpf.rst

```diff
@@ -0,0 +1,307 @@
+  functions above can be called.
+
+* ``test_finish()`` - this function submits the results and returns from the current function.
+
+Function mocking
+----------------
+
+Being able to mock out a function is a great tool to have when creating tests for a number of 
+reasons. You might for example want to tests what happens if a specific function returns an error 
+to see if it is handled gracefully. You might want to proxy function calls to record if the function
+under test actually called specific dependencies. Or you might want to test code that uses helpers
+which rely on state we can't set in BPF like the routing table.
+
+Mocking is a fairy easy:
+
+1. Create a macro with the exact same name and make it equal to another function name.
+
+2. Create a function with a unique name and the same signature as the function it is replacing.
+
+3. Include the file which contains the definition we are replacing.
+
+The following example mocks out the fib_lookup helper call and replaces it with our
+mocked version, since we don't actually have routes for the IPs we want to test:
+
+.. code-block:: c
+
+    #include "common.h"
+
+    #include "bpf/ctx/xdp.h"
+
+    #define fib_lookup mock_fib_lookup
+
+    static const char fib_smac[6] = {0xDE, 0xAD, 0xBE, 0xEF, 0x01, 0x02};
+    static const char fib_dmac[6] = {0x13, 0x37, 0x13, 0x37, 0x13, 0x37};
+
+    long mock_fib_lookup(__maybe_unused void *ctx, struct bpf_fib_lookup *params,
+                __maybe_unused int plen, __maybe_unused __u32 flags)
+    {
+        memcpy(params->smac, fib_smac, sizeof(fib_smac));
+        memcpy(params->dmac, fib_dmac, sizeof(fib_dmac));
+        return 0;
+    }
+
+    #include "bpf_xdp.c"
+    #include "lib/nodeport.h"
+
+Limitations
+-----------
+
+For all its benefits there are some limitations to this way of testing:
+
+* Code must pass the verifier, so our setup and test code has to obey the same rules as other BPF
+  programs. A side effect is that it automatically guarantees that all code that passes will also
+  load. The biggest concern is the complexity limit on older kernels, this can be somewhat mitigated
+  by separating heavy setup work into its own ``SETUP`` program and optionally tail calling into the 
+  to be tested code to ensure the testing harness doesn't push us over the complexity limit.
+
+* Test functions like ``test_log()``, ``test_fail()``, ``test_skip()`` can only be executed within the 
+  scope of the main program or a ``TEST``. These function rely on variables set by ``test_init()`` and
+  will produce errors when used in other functions. 
+  
+* Functions that stop execution(``test_fail_now()``, ``test_fatal()``, ``test_skip_now()``) can't be
+  used within a sub-test(``TEST``) and ``for``, ``while``, or ``select`` since it used ``break`` to stop a
+  sub-test. These functions can still be used from within ``for``, ``while`` and ``select`` if no 
+  sub-tests are used since here the flow interruption happens via ``return``.
```

Comment:
```suggestion
* Functions that halt the execution (``test_fail_now()``, ``test_fatal()``, ``test_skip_now()``) can't be
  used within a sub-test (``TEST``) and ``for``, ``while``, or ``switch/case`` blocks since they use the ``break`` keyword to stop a
  sub-test. These functions can still be used from within ``for``, ``while`` and ``switch/case`` blocks if no 
  sub-tests are used, because in that case the flow interruption happens via ``return``.
```